### PR TITLE
Implement oauth secret token in headful-lite.

### PR DESCRIPTION
**Pull request by Yue-Wang-Google**
Also deprecate TwitterAuth{Provider,Credential}. OAuth sign in flow is now the recommended way to sign in with Twitter.

### CODE REVIEW

**Comment by Yue-Wang-Google** [comment]
DO_NOT_SUBMIT. 

This is for Chuan to review first. I'll wait until Leo's documentation is out.

**Comment by Yue-Wang-Google** [comment]
Relevant internal bug: b/129413855

File: Firebase/Auth/Source/FIRAuth.m [hunk 1]
```diff
@@ -857,12 +857,6 @@ - (void)internalSignInAndRetrieveDataWithCredential:(FIRAuthCredential *)credent
                                        requestConfiguration:_requestConfiguration];
   request.autoCreate = !isReauthentication;
   [credential prepareVerifyAssertionRequest:request];
-  if ([credential isKindOfClass:[FIROAuthCredential class]]) {
```
Comment (Yue-Wang-Google) [inline Firebase/Auth/Source/FIRAuth.m]: just to let reviewer know --- this is moved to FIROAuthCredential's prepareVerifyAssertionRequest:

File: Firebase/Auth/Source/AuthProviders/OAuth/FIROAuthCredential.m [hunk 2]
```diff
@@ -40,30 +41,50 @@ - (nullable instancetype)initWithProvider:(NSString *)provider {
...
+      [self initWithProviderID:providerID IDToken:nil accessToken:nil pendingToken:nil secret:nil];
   if (self) {
     _OAuthResponseURLString = OAuthResponseURLString;
     _sessionID = sessionID;
   }
   return self;
 }
 
+
+- (nullable instancetype)initWithVerifyAssertionResponse:(FIRVerifyAssertionResponse *)response {
```
Comment (Yue-Wang-Google) [inline Firebase/Auth/Source/AuthProviders/OAuth/FIROAuthCredential.m]: there're various cases we need to generate a OAuthCredential from verifyAssertion. So we create a method here.

File: Firebase/Auth/Source/Public/FIRAuthDataResult.h [hunk 3]
```diff
@@ -37,7 +38,14 @@ NS_SWIFT_NAME(AuthDataResult)
...
     @brief The signed in user.
  */
-@property(nonatomic, readonly) FIRUser *user;
+@property(nonatomic, readonly, nullable) FIRUser *user;
+
+/** @property credential
+    @brief The updated OAuth credential after the the sign-in, link and reauthenticate action.
+    @detial This property is for OAuth sign in only.
+ */
+@property(nonatomic, readonly, nullable) FIROAuthCredential *credential;
```
Comment (Yue-Wang-Google) [inline Firebase/Auth/Source/Public/FIRAuthDataResult.h]: This API was approved and both Web and Android has it, but somehow iOS missed it before I work here.
So I added it

File: Firebase/Auth/Source/FIRAuthErrorUtils.h [hunk 4]
```diff
@@ -219,10 +219,12 @@ NS_ASSUME_NONNULL_BEGIN
 /** @fn accountExistsWithDifferentCredentialErrorWithEmail:
     @brief Constructs an @c NSError with the @c FIRAuthErrorAccountExistsWithDifferentCredential
         code.
-    @param Email The email address that is already associated with an existing account
+    @param email The email address that is already associated with an existing account
+    @param updatedCredential The updated credential for the existing account
     @return The NSError instance associated with the given FIRAuthError.
  */
-+ (NSError *)accountExistsWithDifferentCredentialErrorWithEmail:(nullable NSString *)Email;
++ (NSError *)accountExistsWithDifferentCredentialErrorWithEmail:(nullable NSString *)email
```
Comment (Yue-Wang-Google) [inline Firebase/Auth/Source/FIRAuthErrorUtils.h]: we need to return the error with updated credential in userInfo. so added a new param.

File: Firebase/Auth/Source/AuthProviders/OAuth/FIROAuthCredential_Internal.h [hunk 5]
```diff
@@ -46,11 +48,13 @@ NS_ASSUME_NONNULL_BEGIN
     @param IDToken The ID Token associated with the credential being created.
     @param accessToken The access token associated with the credential being created.
     @param pendingToken The pending token associated with the credential being created.
+    @param secret The secret associated with the credential being created.
  */
 - (instancetype)initWithProviderID:(NSString *)providerID
                            IDToken:(nullable NSString *)IDToken
                        accessToken:(nullable NSString *)accessToken
-                      pendingToken:(nullable NSString *)pendingToken NS_DESIGNATED_INITIALIZER;
+                      pendingToken:(nullable NSString *)pendingToken
```
Comment (renkelvin) [inline Firebase/Auth/Source/AuthProviders/OAuth/FIROAuthCredential_Internal.h]: I prefer switching the order of pendingToken and secret.

Comment (Yue-Wang-Google) [reply Firebase/Auth/Source/AuthProviders/OAuth/FIROAuthCredential_Internal.h]: Done.

File: Firebase/Auth/Source/FIRUser.m [hunk 6]
```diff
@@ -18,6 +18,7 @@
 
 #import <FirebaseCore/FIRLogger.h>
 
+#import "AuthProviders/OAuth/FIROAuthCredential_Internal.h"
```
[the remark this answers is not in the document]
Comment (Yue-Wang-Google) [reply Firebase/Auth/Source/FIRUser.m]: Done

File: Example/Auth/Tests/FIRTwitterAuthProviderTests.m [hunk 7]
```diff
@@ -47,6 +47,8 @@ @implementation FIRTwitterAuthProviderTests
         the appropriate fields in a verify assertion request.
  */
 - (void)testCredentialWithToken {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
```
[the remark this answers is not in the document]
Comment (Yue-Wang-Google) [reply Example/Auth/Tests/FIRTwitterAuthProviderTests.m]: Done.

**Comment by Yue-Wang-Google** [comment]
Doc just got approved. Submitting.

**Pull request merged**